coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+			}
+
+			err = reportFailedWorkspaceBuilds(ctx, logger, db, enqueuer, clk)
+			if err != nil {
+				logger.Debug(ctx, "unable to report failed workspace builds")
+				return err
+			}
+
+			logger.Info(ctx, "report generator finished", slog.F("duration", clk.Since(start)))
+
+			return nil
+		}, nil); err != nil {
+			logger.Error(ctx, "failed to generate reports", slog.Error(err))
+			return
+		}
+	}
+
+	go func() {
+		defer close(closed)
+		defer ticker.Stop()
+		// Force an initial tick.
+		doTick(dbtime.Time(clk.Now()).UTC())
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case tick := <-ticker.C:
+				ticker.Stop()
+				doTick(dbtime.Time(tick).UTC())
+			}
+		}
+	}()
+	return &reportGenerator{
+		cancel: cancelFunc,
+		closed: closed,
+	}
+}
+
+type reportGenerator struct {
+	cancel context.CancelFunc
+	closed chan struct{}
+}
+
+func (i *reportGenerator) Close() error {
+	i.cancel()
+	<-i.closed
+	return nil
+}
+
+const failedWorkspaceBuildsReportFrequencyDays = 7
+
+func reportFailedWorkspaceBuilds(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) error {
+	now := clk.Now()
+	since := now.Add(-failedWorkspaceBuildsReportFrequencyDays * 24 * time.Hour)
+
+	statsRows, err := db.GetWorkspaceBuildStatsByTemplates(ctx, dbtime.Time(since).UTC())
+	if err != nil {
+		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
+	}
+	sort.Slice(statsRows, func(i, j int) bool {
+		return statsRows[i].TemplateName < statsRows[j].TemplateName
+	})
+
+	reportGeneratedNow := map[uuid.UUID]bool{}
+	for _, stats := range statsRows {
+		var failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow
+		reportData := map[string]any{}
+
+		if stats.FailedBuilds > 0 {
+			failedBuilds, err = db.GetFailedWorkspaceBuildsByTemplateID(ctx, database.GetFailedWorkspaceBuildsByTemplateIDParams{
+				TemplateID: stats.TemplateID,
+				Since:      dbtime.Time(since).UTC(),
+			})
+			if err != nil {
+				logger.Error(ctx, "unable to fetch failed workspace builds", slog.F("template_id", stats.TemplateID), slog.Error(err))
+				continue
+			}
+
+			// There are some failed builds, so we have to prepare input data for the report.
+			reportData = buildDataForReportFailedWorkspaceBuilds(failedWorkspaceBuildsReportFrequencyDays, stats, failedBuilds)
+		}
+
+		templateAdmins, err := findTemplateAdmins(ctx, db, stats)
+		if err != nil {
+			logger.Error(ctx, "unable to find template admins", slog.F("template_id", stats.TemplateID), slog.Error(err))
+			continue
+		}
+
+		for _, templateAdmin := range templateAdmins {
+			reportLog, err := db.GetReportGeneratorLogByUserAndTemplate(ctx, database.GetReportGeneratorLogByUserAndTemplateParams{
+				UserID:                 templateAdmin.ID,
+				NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			})
+			if err != nil && !xerrors.Is(err, sql.ErrNoRows) { // sql.ErrNoRows: report not generated yet
```

Comment:
Under what circumstances will this occur?

Reply:
What do you mean? A generic error or `sql.ErrNoRows`?

Re `sql.ErrNoRows`: First generator run for a user. No log entries have been stored yet.

Reply:
> Re sql.ErrNoRows: First generator run for a user. No log entries have been stored yet.

Is that an error case?

Reply:
Removed `logger.Error`.